build_resolvers/lib/src/analysis_driver_filesystem.dart

```diff
@@ -0,0 +1,173 @@
+// Copyright (c) 2025, the Dart project authors. Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+import 'dart:convert';
+import 'dart:typed_data';
+
+import 'package:analyzer/file_system/file_system.dart';
+import 'package:analyzer/source/file_source.dart';
+// ignore: implementation_imports
+import 'package:analyzer/src/clients/build_resolvers/build_resolvers.dart';
+import 'package:build/build.dart' hide Resource;
+import 'package:path/path.dart' as p;
+
+/// The in-memory filesystem that is the analyzer's view of the build.
+///
+/// Tracks modified paths, which should be passed to
+/// `AnalysisDriver.changeFile` to update the analyzer state.
+class AnalysisDriverFilesystem implements UriResolver, ResourceProvider {
+  final Map<String, String> _data = {};
+  final Set<String> _changedPaths = {};
+
+  // Methods for use by `AnalysisDriverModel`.
+
+  /// Whether [path] exists.
+  bool exists(String path) => _data.containsKey(path);
+
+  /// Reads the data previously written to [path].
+  ///
+  /// Throws if ![exists].
+  String read(String path) => _data[path]!;
+
+  /// Deletes the data previously written to [path].
+  ///
+  /// Records the change in [changedPaths].
+  ///
+  /// Or, if it's missing, does nothing.
+  void deleteFile(String path) {
+    if (_data.remove(path) != null) _changedPaths.add(path);
+  }
+
+  /// Writes [content] to [path].
+  ///
+  /// Records the change in [changedPaths], only if the content actually
+  /// changed.
+  void writeFile(String path, String content) {
+    final oldContent = _data[path];
+    _data[path] = content;
+    if (content != oldContent) _changedPaths.add(path);
+  }
+
+  /// Paths that were modified by [deleteFile] or [writeFile] since the last
+  /// call to [clearChangedPaths].
+  Iterable<String> get changedPaths => _changedPaths;
+
+  /// Clears [changedPaths].
+  void clearChangedPaths() => _changedPaths.clear();
+
+  // `UriResolver` methods.
+
+  @override
+  Uri pathToUri(String path) {
```

Comment:
What is the first segment of the path that we are always skipping?

It might be good to put a comment with the expected path format here.